sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java

```diff
@@ -891,7 +899,7 @@ private Object createCacheKey(
         // For caching reader, we don't care about the watermark.
         return restrictionCoder.structuralValue(
             UnboundedSourceRestriction.create(
-                source, checkpoint, BoundedWindow.TIMESTAMP_MIN_VALUE));
+                source, checkpoint, BoundedWindow.TIMESTAMP_MIN_VALUE, null));
       }
```

Comment:
Depending on whether source splits are kept locally or repeatedly sent over the data API you might want to use the split id instead of the instance as a key to avoid issues with source implementations that don't override `#equals(Object)` and `#hashCode()`.